### ISSUE TITLE
Add Maven Support 

### ISSUE THREAD

**Issue by danyaljj** [google/or-tools#202]
Original post
I can see some maven releases here: 
https://bintray.com/nikola-ilo/maven/or-tools/view
Q: Why the releases are not continued? Are you guys planning to release a new version on maven? 
A: Not done by Google LLC

For v8.0 we are migrating to use maven to create java package
Todo List:
* [x] Use Maven in CMake based build
  * [x] Add Maven package to CMake based build
  * [x] Use Maven for samples (`ortools/*/samples`)
  * [x] Use Maven for examples  (`examples/contrib examples/java`)
  * [x] Use Maven for tests (`examples/tests`)
* [x] Use Maven in Make based build
  * [x] Add Maven packages to Make based build
  * [x] Use Maven for samples
  * [x] Use Maven for examples
  * [x] Use Maven for tests
* [x] Rework archive to use Maven packages
  * [x] Linux archive
  * [x] MacOS archive
  * [x] Windows archive
* [x] Add internal doc in:
  * [x] ortools/java/README.md
  * [x] ortools/java/doc/local_pipeline.dot
  * [x] cmake/doc/java.md

**Comment by lperron** [comment]
These are actual contributed and we did not do it.
I will contact the author to see if we can transfer it to or-tools.

Thanks

Laurent Perron | Operations Research | lperron@google.com | (33) 1 42 68 53
00


2016-06-30 0:20 GMT+02:00 Daniel Khashabi <notifications@github.com>:

**Comment by PunKeel** [comment]
I would definitely appreciate being able to integrate or-tools using Maven

**Comment by dlahlou** [comment]
Hello Punkeel,
It's on our to-do list to support Maven. Thank you.
Regards,
Driss

On Aug 26, 2016 03:36, "PunKeel" <notifications@github.com> wrote:

I would definitely appreciate being able to integrate or-tools using Maven

—
You are receiving this because you are subscribed to this thread.
Reply to this email directly, view it on GitHub
<https://github.com/google/or-tools/issues/202#issuecomment-242597370>, or mute
the thread

**Comment by guizmaii** [comment]
Any news ?

**Comment by fcorneli** [comment]
Maven support would be interesting...

**Comment by Mizux** [comment]
hi all,

Yes I would like to provide a maven (gradle ? ivy ?) package for or-tools ASAP and ideas are welcome !

How I see it currently:
1.Have a clean and modern CMake project to replace Makefile (cf WIP mizux/cmake branch)
2. Have a gradle project with cpp-plugin/JNI to wrap libortools.so
3. Use gradle plugin "maven-publish'
4. Profit !

**Comment by gsavvid** [comment]
Hello, 

I was wondering whether this feature is still planned to be included in v6.8 and also whether the due date of v6.8 (April 1st) is still valid.

**Comment by Mizux** [comment]
Hi,

Not sure it will be included in v6.8 and not sure when v6.8 will be out (more mid-april I think).
The Goal is to have a Release (which means a version (aka SHA1) generated on Linux/Mac/Win) every month if possible.
For features, it's a best effort basis and depends to the bugs/features found by the community.

So for Maven, I would say, hope to do it ASAP (so I put it in my kaban/ToDo list) but I can't really give you a strong/mandatory date...

**Comment by oliviercailloux** [comment]
The project is now published to Maven Central

**Comment by Mizux** [comment]
FYI: rpath should be fixed now on unix (linux and mac), (win use static deps -> not applicable)

```bash
[0]─[~/work/or-tools]
[^v^]─corentinl@corentinl %ldd lib/libjniortools.so               
	linux-vdso.so.1 (0x00007ffc285fe000)
	libgflags.so.2.2 => /usr/local/google/home/corentinl/work/or-tools/lib/../dependencies/install/lib/libgflags.so.2.2 (0x00007f4e30f3e000)
	libglog.so.0.3.5 => /usr/local/google/home/corentinl/work/or-tools/lib/../dependencies/install/lib/libglog.so.0.3.5 (0x00007f4e30d03000)
	libprotobuf.so.3.5.1 => /usr/local/google/home/corentinl/work/or-tools/lib/../dependencies/install/lib/libprotobuf.so.3.5.1 (0x00007f4e304dc000)
	libCbcSolver.so.3 => /usr/local/google/home/corentinl/work/or-tools/lib/../dependencies/install/lib/libCbcSolver.so.3 (0x00007f4e301e5000)
	libCbc.so.3 => /usr/local/google/home/corentinl/work/or-tools/lib/../dependencies/install/lib/libCbc.so.3 (0x00007f4e2fec7000)
	libOsiCbc.so.3 => /usr/local/google/home/corentinl/work/or-tools/lib/../dependencies/install/lib/libOsiCbc.so.3 (0x00007f4e2fcb7000)
	libCgl.so.1 => /usr/local/google/home/corentinl/work/or-tools/lib/../dependencies/install/lib/libCgl.so.1 (0x00007f4e2f999000)
	libClpSolver.so.1 => /usr/local/google/home/corentinl/work/or-tools/lib/../dependencies/install/lib/libClpSolver.so.1 (0x00007f4e2f73d000)
	libClp.so.1 => /usr/local/google/home/corentinl/work/or-tools/lib/../dependencies/install/lib/libClp.so.1 (0x00007f4e2f38c000)
	libOsiClp.so.1 => /usr/local/google/home/corentinl/work/or-tools/lib/../dependencies/install/lib/libOsiClp.so.1 (0x00007f4e2f141000)
	libOsi.so.1 => /usr/local/google/home/corentinl/work/or-tools/lib/../dependencies/install/lib/libOsi.so.1 (0x00007f4e2eeee000)
	libCoinUtils.so.3 => /usr/local/google/home/corentinl/work/or-tools/lib/../dependencies/install/lib/libCoinUtils.so.3 (0x00007f4e2eb9b000)
	libortools.so => /usr/local/google/home/corentinl/work/or-tools/lib/libortools.so (0x00007f4e2e0f7000)
	libz.so.1 => /lib/x86_64-linux-gnu/libz.so.1 (0x00007f4e2dedd000)
	librt.so.1 => /lib/x86_64-linux-gnu/librt.so.1 (0x00007f4e2dcd5000)
	libpthread.so.0 => /lib/x86_64-linux-gnu/libpthread.so.0 (0x00007f4e2dab8000)
	libstdc++.so.6 => /usr/lib/x86_64-linux-gnu/libstdc++.so.6 (0x00007f4e2d733000)
	libm.so.6 => /lib/x86_64-linux-gnu/libm.so.6 (0x00007f4e2d42f000)
	libgcc_s.so.1 => /lib/x86_64-linux-gnu/libgcc_s.so.1 (0x00007f4e2d217000)
	libc.so.6 => /lib/x86_64-linux-gnu/libc.so.6 (0x00007f4e2ce78000)
	/lib64/ld-linux-x86-64.so.2 (0x00007f4e314a8000)
[0]─[~/work/or-tools]
[^v^]─corentinl@corentinl %objdump -p lib/libjniortools.so | grep RUNPATH
  RUNPATH              $ORIGIN:$ORIGIN/../lib:$ORIGIN/../dependencies/install/lib64:$ORIGIN/../dependencies/install/lib
```

**Comment by gsavvid** [comment]
Until this feature is released, if anyone is interested in building OR-tools with Gradle here's an example of how I did it https://stackoverflow.com/a/52790798/1680301.

**Comment by Vamsi-Legion** [comment]
Any official maven version ?

**Comment by magneticflux-** [comment]
I think I have a robust solution for consumers in my repository [here](https://github.com/magneticflux-/ortools-java). I have dependencies available for each distribution of OR-Tools, and a main dependency that extracts the binaries at runtime. I believe this is similar to how the .NET publication works currently.

**Comment by rsahani** [comment]
Hi, Is there a tentative timeline for this release?

**Comment by Mizux** [comment]
Using the CMake based build system you can already generate them.
For next release I would like to integrate it to our makefile based build.

**Comment by dillonmabry** [comment]
Will this ever be publicly available as a simple maven/gradle repo? Not sure why it isn't....

**Comment by amucunguzi** [comment]
If anyone is interested in a maven dependency for the latest version (7.8), our team has created one [here](https://github.com/panavis/google-or-tools-maven).

**Comment by magneticflux-** [comment]
My solution from above https://github.com/google/or-tools/issues/202#issuecomment-614399857 is updated to 7.8 and published to JCenter. The readme of the project explains how it differs from existing solutions.

**Comment by oliviercailloux** [comment]
> If anyone is interested in a maven dependency for the latest version (7.8), our team has generated one [here](https://github.com/panavis/google-or-tools-maven) and published it on maven central. Also added instructions on how to reproduce the project.
> 
> Was inspired by similar work by @oliviercailloux and @Mizux pointed us in the right direction (used `cmake` instead of `makefile`; and loaded native libraries using the `Loader.java` class).

Thanks for this work, and for the back references. About the link to my pom for publication to Maven central, please note that I have a https://github.com/oliviercailloux/java-course/blob/master/Maven/Maven%20central.adoc[document] about publishing to Maven Central, that may perhaps be useful.

**Comment by lalithsuresh** [comment]
@Mizux will the maven packages for release 8.0 be made available on Maven Central?

**Comment by Mizux** [comment]
This is the next step, but we need to test few things:
* how to sign package while still given the opportunity to build locally without the "official" token
* currently we have three "meta" packages
  * Linux: `ortools-java` depends on `ortools-linux-x86-64`
  * MacOS: `ortools-java` depends on `ortools-darwin`
  * Windows: `ortools-java` depends on `ortools-win32-x86-64`
* I don't know if the code generated for the meta package containing the java class is cross-platform and identical,
if yes then we will be able to merge all `ortools-java` in one package and publish it to maven (ed this is what we do for .NET)
ref: https://github.com/google/or-tools/tree/stable/ortools/java
ref: https://github.com/google/or-tools/tree/stable/ortools/dotnet#complete-googleortools-package

I'll **should** work on it next week

**Comment by dillonmabry** [comment]
@Mizux any update on this or from someone else? I believe quite a few people would be very happy to have this added on maven central for Java since quite a few teams would be using this for optimization problems :)

**Comment by amucunguzi** [comment]
@dillonmabry our team has been using this: https://github.com/panavis/google-or-tools-maven. We hope that there will be an official release soon. @Mizux.

**Comment by suming** [comment]
@amucunguzi thank you for working on this! I saw that https://github.com/panavis/google-or-tools-maven doesn't have a dependency for mac -- is that on it's way at all?

**Comment by magneticflux-** [comment]
@suming My version https://github.com/google/or-tools/issues/202#issuecomment-614399857 has all OS distributions available :eyes:

**Comment by lalithsuresh** [comment]
Yes, I've been using @magneticflux-'s version on OSX and it works fine.

**Comment by suming** [comment]
@magneticflux- & @lalithsuresh thank you for the heads-up, I just tested it out and it's great. I can develop locally on OSX and deploy on Linux easily. Appreciate it!

**Comment by dillonmabry** [comment]
> @suming My version [#202 (comment)](https://github.com/google/or-tools/issues/202#issuecomment-614399857) has all OS distributions available 👀

Using this version currently with no issues, thanks for your support! Works as expected.

**Comment by cbtum** [comment]
> @suming My version [#202 (comment)](https://github.com/google/or-tools/issues/202#issuecomment-614399857) has all OS distributions available 👀

That's the only approach that worked for me on macOs (under maven and gradle). Thank you so much!

After maven is now officially supported in ortools v8, I tried to use it  but gave up after two days (even after a discussion in the google group). 

@magneticflux- , I guess this has to do with how you "handle more complex classpaths"? Any chance your approach will be integrated in google's ortools in the long run so that we can all use the "official" version?

**Comment by ErickPol** [comment]
> @suming My version [#202 (comment)](https://github.com/google/or-tools/issues/202#issuecomment-614399857) has all OS distributions available 👀

For those that had trouble running Makefile in order to install the jars follows a detailed set-up guide based on @magneticflux- 's solution:

1. Guarantee maven (mvn) executable folder (e.g. /bin) is on your classpath
2. Download binaries (version 8.0) for your OS at https://github.com/google/or-tools/releases/tag/v8.0. For Windows the link is https://github.com/google/or-tools/releases/download/v8.0/or-tools_VisualStudio2019-64bit_v8.0.8283.zip
3. Extract the zip file and run these 2 commands inside the folder:
`mvn org.apache.maven.plugins:maven-install-plugin:2.5.2:install-file -Dfile="ortools-win32-x86-64-8.0.8283.jar" -DpomFile="pom-runtime.xml"`
`mvn org.apache.maven.plugins:maven-install-plugin:2.5.2:install-file -Dfile="ortools-java-8.0.8283.jar" -DpomFile="pom-local.xml"`
4. insert the following dependencies on your project (in my case, to run or-tools over windows) :
`        <dependency>
            <groupId>com.skaggsm.ortools</groupId>
            <artifactId>ortools-natives-all</artifactId>
            <version>8.0.8283</version>
            <type>pom</type>
        </dependency>
        <dependency>
            <groupId>com.google.ortools</groupId>
            <artifactId>ortools-java</artifactId>
            <version>8.0.8283</version>
        </dependency>
        <dependency>
            <groupId>com.google.ortools</groupId>
            <artifactId>ortools-win32-x86-64</artifactId>
            <version>8.0.8283</version>
        </dependency>
`
5. create a test file , calling first the loader and then setting up the solver:
`    @Test
    public void test(){
        OrToolsHelper.loadLibrary();
        MPSolver solver = MPSolver.createSolver("SCIP");
    }
`

**Comment by magneticflux-** [comment]
> 
> 
> > @suming My version [#202 (comment)](https://github.com/google/or-tools/issues/202#issuecomment-614399857) has all OS distributions available 👀
> 
> For those that had trouble running Makefile in order to install the jars follows a detailed set-up guide based on @magneticflux- 's solution:
> 
>     1. Guarantee maven (mvn) executable folder (e.g. /bin) is on your classpath
> 
>     2. Download binaries (version 8.0) for your OS at https://github.com/google/or-tools/releases/tag/v8.0. For Windows the link is https://github.com/google/or-tools/releases/download/v8.0/or-tools_VisualStudio2019-64bit_v8.0.8283.zip
> 
>     3. Extract the zip file and run these 2 commands inside the folder:
>        `mvn org.apache.maven.plugins:maven-install-plugin:2.5.2:install-file -Dfile="ortools-win32-x86-64-8.0.8283.jar" -DpomFile="pom-runtime.xml"`
>        `mvn org.apache.maven.plugins:maven-install-plugin:2.5.2:install-file -Dfile="ortools-java-8.0.8283.jar" -DpomFile="pom-local.xml"`
> 
>     4. Add the following dependencies to your project (in my case, to run or-tools over windows) :
>        `       <dependency> <groupId>com.skaggsm.ortools</groupId> <artifactId>ortools-natives-all</artifactId> <version>8.0.8283</version> <type>pom</type> </dependency> <dependency> <groupId>com.google.ortools</groupId> <artifactId>ortools-java</artifactId> <version>8.0.8283</version> </dependency> <dependency> <groupId>com.google.ortools</groupId> <artifactId>ortools-win32-x86-64</artifactId> <version>8.0.8283</version> </dependency>`
> 
>     5. Run a test file , calling first the loader and then setting up the solver:
>        `   @Test public void test(){ OrToolsHelper.loadLibrary(); MPSolver solver = MPSolver.createSolver("SCIP"); }`

I'm not sure why you're installing both my distribution and the official one. You don't need both; mine has everything included and has extra extraction functionality.

**Comment by jbellis** [comment]
> This is the next step, but we need to test few things:

Is there a ticket for maven central publishing?

**Comment by magneticflux-** [comment]
> Is there a ticket for maven central publishing?

With the sunsetting of JCenter, this is more urgent for me in particular. I'm not keen on shouldering the cost of hosting large binaries myself going forward, so I'd like my solution to just depend on official binaries on Maven Central.

**Comment by Mizux** [comment]
First 8.2 RC (8.2.9004)
https://search.maven.org/search?q=g:com.google.ortools
https://mvnrepository.com/artifact/com.google.ortools

note: I failed the linux artifact which is currently depending on GLIBC 2.33

**Comment by lalithsuresh** [comment]
@Mizux thanks! I just tested the Darwin version and it works.

Could you elaborate more on the Linux artifact? 

If it helps, I'm getting an error about needing GLIBC 2.29 when running on Ubuntu 18.04.

```
java.lang.UnsatisfiedLinkError: /tmp/ortools-java7239044082451951181/linux-x86-64/libjniortools.so: /lib/x86_64-linux-gnu/libm.so.6: version `GLIBC_2.29' not found (required by /tmp/ortools-java7239044082451951181/linux-x86-64/libortools.so)
```

**Comment by Mizux** [comment]
see https://github.com/google/or-tools/issues/2520
TLDR the runtime has been built against GLIBC 33 (archlinux) need to build the package against an "older" Linux distro

**Comment by dillonmabry** [comment]
> see #2520
> TLDR the runtime has been built against GLIBC 33 (archlinux) need to build the package against an "older" Linux distro

This is great news, and we appreciate your hard work!

Question: Is there a solution to have this work with Docker? For example, it works fine for me locally but whenever I run my docker build it cannot find the JNI/Library resource files like it can when I run locally, it extracts and discovers them automatically with my Spring/Java build.

I am using JDK 11 and Maven if that helps.

Thank you!

**Comment by Mizux** [comment]
@dillonmabry let's open a new issue with your docker issue.
note: You are not the only one with this feedback IIRC but I was never be able to reproduce it...
see: https://github.com/or-tools/java_or-tools/blob/main/Dockerfile and https://github.com/or-tools/java_or-tools/runs/2441985470?check_suite_focus=true
